### PR TITLE
Update Callback subclass in callbacks.py

### PR DESCRIPTION
**Pull request by 8bitmp3**
See #29958 - Suggested improvements:

- _Expanded the description:_
To make things clear, added more information so that a user knows they can use `Callback` as a custom callback during training, evaluating and inference when the existing built-in callback is not enough.

- _Added a `Callback` example:_
This callback doc needed an example. It was adapted/modified from the Write Custom Callbacks [guide](https://www.tensorflow.org/beta/guide/keras/custom_callback) (r2.0b). Please note that `steps_per_epoch` in the `.fit()` method during training may be GPU memory hungry, so a larger batch size was used (128) on a training set (55000, Fashion MNIST) to mitigate it when testing this custom callback example.

### CODE REVIEW

File: tensorflow/python/keras/callbacks.py [hunk 1]
```diff
@@ -443,6 +449,38 @@ class Callback(object):
...
           (if accuracy monitoring is enabled).
+    
+  Example:
+  ```python  
+  # Import the datetime module to display the time
+  import datetime
+  
+  # A custom callback displaying the loss and accuracy 
+  # from the `logs` dict at the end of a batch/epoch
+  class MyLossAccuracyCallback(tf.keras.callbacks.Callback):
```
Comment (rchao) [inline tensorflow/python/keras/callbacks.py]: https://www.tensorflow.org/beta/guide/keras/custom_callback has a similar example. Can we make the input, code and class name in this example be consistent with that?

Comment (8bitmp3) [reply tensorflow/python/keras/callbacks.py]: @rchao Absolutely. See below. Thank you!

Small changes to that example from the Keras Custom Callback in TF 2 Beta [guide](https://www.tensorflow.org/beta/guide/keras/custom_callback): 

1) `x_train` and `y_train` -> `data` and `labels` for consistency, so it's just like most callback examples in `callbacks.py`, unless you want to use the "classic" `x` and `y` notation. 

2) Also, it's probably good to still define `steps_per_epoch` - as before - as an iteration over training data i.e. length of training data (`x_train` or `data`) divided by `batch_size` (See this [TF 2 Beta tutorial of TF Hub with Keras](https://www.tensorflow.org/beta/tutorials/images/hub_with_keras) which uses the same approach).

```python
# A custom callback displaying the loss and mean absolute error 
# from the `logs` dict at the end of a batch/epoch
class LossAndErrorPrintingCallback(tf.keras.callbacks.Callback):

  def on_train_batch_end(self, batch, logs=None):
    print('For batch {}, loss is {:7.2f}.'.format(batch, logs['loss']))

  def on_test_batch_end(self, batch, logs=None):
    print('For batch {}, loss is {:7.2f}.'.format(batch, logs['loss']))

  def on_epoch_end(self, epoch, logs=None):
    print('The average loss for epoch {} is {:7.2f} and mean absolute error is {:7.2f}.'.format(epoch, logs['loss'], logs['mae']))

batch_size=64
epochs=3

model.fit(data, labels,
          batch_size=batch_size,
          steps_per_epoch=int(np.ceil(data.shape[0])/batch_size),
          epochs=epochs,
          verbose=0,
          callbacks=[LossAndErrorPrintingCallback()])
```

Comment (8bitmp3) [reply tensorflow/python/keras/callbacks.py]: @rchao @rthadur if there are further improvements needed to be done let me know. thanks for the help 👍

File: tensorflow/python/keras/callbacks.py [hunk 2]
```diff
@@ -443,6 +449,38 @@ class Callback(object):
...
+    def on_epoch_end(self, epoch, logs=None):
+      print('Average loss for epoch {} = {:7.2f}, accuracy = {:7.2f}.'
+            .format(epoch, logs['loss'], logs['acc']))
+  
+  batch_size = 128
+  epochs = 10
+  
+  model.fit(data, labels, 
+            batch_size=batch_size,
+            steps_per_epoch=int(np.ceil(X_train.shape[0])/batch_size),
```
Comment (rchao) [inline tensorflow/python/keras/callbacks.py]: By x_train did you mean data?
Also, curious, what is the reason why we need np.ceil here?

**Review by rchao — COMMENTED** [review]
Apologies for the wait. Thanks for making the changes!

**Review by lamberta — DISMISSED** [review]
Doc review looks good to me, thanks!

File: tensorflow/python/keras/callbacks.py (hunk 2 above)
Comment (8bitmp3) [reply tensorflow/python/keras/callbacks.py]: Just saw this @rchao:

1) Replacing `X_train` to `data` - habits!

2) Used `np.ceil` to be in line with @fchollet 's example [here](https://keras.io/examples/mnist_tfrecord/) on www.keras.io:
```python
train_model.fit(
    epochs=epochs,
    steps_per_epoch=int(np.ceil(data.train.num_examples / float(batch_size))),
    callbacks=[EvaluateInputTensor(test_model, steps=100)])
```

**Comment by 8bitmp3** [comment]
> Doc review looks good to me, thanks!

@lamberta can we undo "ready to pull" if it needs to be done, fixed "X_train" -> "data" thanks to @rchao 's eagle eye 👍 as per his earlier comment

File: tensorflow/python/keras/callbacks.py [hunk 3]
```diff
@@ -443,6 +449,38 @@ class Callback(object):
...
+  ```python  
+  # Import the datetime module to display the time
+  import datetime
+  
+  # A custom callback displaying the loss and accuracy 
+  # from the `logs` dict at the end of a batch/epoch
+  class MyLossAccuracyCallback(tf.keras.callbacks.Callback):
+  
+    def on_train_batch_end(self, batch, logs=None):
+      print('For batch {}, loss = {:7.2f} at '.format(batch, logs['loss']),
```
Comment (rthadur) [inline tensorflow/python/keras/callbacks.py]: print statement is falling internal checks , can you please remove, thank you

File: tensorflow/python/keras/callbacks.py [hunk 4]
```diff
@@ -443,6 +449,38 @@ class Callback(object):
...
+  # A custom callback displaying the loss and accuracy 
+  # from the `logs` dict at the end of a batch/epoch
+  class MyLossAccuracyCallback(tf.keras.callbacks.Callback):
+  
+    def on_train_batch_end(self, batch, logs=None):
+      print('For batch {}, loss = {:7.2f} at '.format(batch, logs['loss']), 
+            datetime.datetime.now().time())
+    
+    def on_test_batch_end(self, batch, logs=None):
+      print('For batch {}, loss = {:7.2f} at '.format(batch, logs['loss']),
```
Comment (rthadur) [inline tensorflow/python/keras/callbacks.py]: same as above

File: tensorflow/python/keras/callbacks.py [hunk 5]
```diff
@@ -443,6 +449,38 @@ class Callback(object):
...
+            datetime.datetime.now().time())
+            
+    def on_epoch_end(self, epoch, logs=None):
+      print('Average loss for epoch {} = {:7.2f}, accuracy = {:7.2f}.'
```
Comment (rthadur) [inline tensorflow/python/keras/callbacks.py]: same as above

**Review by rthadur — CHANGES_REQUESTED** [review]

**Comment by mellanox-github** [comment]
Can one of the admins verify this patch?

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by rthadur** [comment]
I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen!
